apps/api/migrations/preference-centralization/preference-centralization.migration.ts

```diff
@@ -0,0 +1,154 @@
+/* eslint-disable no-console */
+import '../../src/config';
+
+import { NestFactory } from '@nestjs/core';
+import { SubscriberPreferenceRepository, NotificationTemplateRepository, PreferenceLevelEnum } from '@novu/dal';
+import {
+  UpsertPreferences,
+  UpsertWorkflowPreferencesCommand,
+  UpsertUserWorkflowPreferencesCommand,
+  UpsertSubscriberGlobalPreferencesCommand,
+  UpsertSubscriberWorkflowPreferencesCommand,
+} from '@novu/application-generic';
+import { buildWorkflowPreferencesFromPreferenceChannels, DEFAULT_WORKFLOW_PREFERENCES } from '@novu/shared';
+
+import { AppModule } from '../../src/app.module';
+
+const BATCH_SIZE = 100;
+
+const sleep = (ms: number) => {
+  return new Promise((resolve) => {
+    setTimeout(resolve, ms);
+  });
+};
+
+/**
+ * Migration to centralize workflow and subscriber preferences.
+ * - subscriber global preference
+ *    -> preferences with subscriber global type
+ * - subscriber workflow preferences
+ *    -> preferences with subscriber workflow type
+ * - workflow preferences
+ *   -> preferences with workflow-resource type
+ *   -> preferences with user-workflow type
+ */
+export async function preferenceCentralization() {
+  const app = await NestFactory.create(AppModule, {
+    logger: false,
+  });
+  console.log('start migration - preference centralization');
+
+  const upsertPreferences = app.get(UpsertPreferences);
+
+  const subscriberPreferenceRepository = app.get(SubscriberPreferenceRepository);
+  const subscriberPreferenceCursor = await subscriberPreferenceRepository._model
```

Comment:
Nitpick: I'd move this closer to the iteration for/loop and instantiate after the workflow preference migration is done.

Reply:
Good call, I separated into 2 functions that are called sequentially for each of Workflows and Subscribers, the cursor is instantiated at the beginning of each.